src/utils/communication_buffer.hpp

```diff
@@ -202,30 +230,32 @@ CommBuffer<T> &CommBuffer<T>::operator=(const CommBuffer<U> &in) {
   comm_ = in.comm_;
   active_ = in.active_;
   my_rank = Globals::my_rank;
+  get_resource_ = in.get_resource_;
   return *this;
 }
 
 template <class T>
-void CommBuffer<T>::Send() noexcept {
+void CommBuffer<T>::Send(bool local, int size) noexcept {
   if (!active_) {
-    SendNull();
+    SendNull(local);
     return;
   }
 
   PARTHENON_DEBUG_REQUIRE(*state_ == BufferState::stale,
                           "Trying to send from buffer that hasn't been staled.");
   *state_ = BufferState::sending;
-  if (*comm_type_ == BuffCommType::sender) {
+  if (*comm_type_ == BuffCommType::sender && !local) {
 // Make sure that this request isn't still out,
 // this could be blocking
 #ifdef MPI_PARALLEL
     PARTHENON_REQUIRE(
         buf_.size() > 0,
         "Trying to send zero size buffer, which will be interpreted as sending_null.");
+    if (size < 0) size = buf_.size();
```

Comment:
Is this check required because of the `-1` default for `size` (so legacy compatibility)?

Reply:
yeah, the basic idea is that `-1` for size implies that the underlying buffer doesn't allow for different sizes (e.g. the buffers are being pulled from a buffer pool rather than being allocated on the fly).